test/e2e/retry.go

```diff
@@ -35,18 +40,40 @@ const (
 
 // retryWithTimeout retries a function that returns an error until a timeout is reached
 func retryWithTimeout(interval, timeout time.Duration, fn func() error) error {
-	var pollError error
-	err := wait.PollImmediate(interval, timeout, func() (bool, error) {
-		pollError = nil
-		err := fn()
-		if err != nil {
-			pollError = err
-			return false, nil //nolint:nilerr // We don't want to return err here
+	var pollError, returnError error
+	result := make(chan error)
+
+	ctx, cancelFunc := context.WithTimeout(context.Background(), timeout)
+	defer cancelFunc()
+
+	go func(ctx context.Context) {
+		result <- wait.PollImmediateWithContext(ctx, interval, timeout, func(ctx context.Context) (bool, error) {
+			if ctx.Err() != nil {
+				return false, ctx.Err()
+			}
+			err := fn() // TODO: mitigate function leaking; fn() can run in the background
```

Comment:
I don't think it's possible to work around this without a way to forcefully kill one goroutine from another goroutine, which seems generally dangerous to do anyway.

Reply:
I agree. Unless `fn()` also watches the passed context.